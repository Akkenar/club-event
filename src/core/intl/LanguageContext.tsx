import * as React from 'react';
import { IntlContext } from './intl.type';
import { DEFAULT_LANGUAGE } from './provideLanguageInRoute';

const LanguageContext = React.createContext<IntlContext>({
  language: DEFAULT_LANGUAGE,
  messages: {},
});

export default LanguageContext;
