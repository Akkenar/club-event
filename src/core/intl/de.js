import messages from '../../../translations/de/messages.de.json';
import confirmation from '../../../translations/de/confirmation.de.md';
import information from '../../../translations/de/information.de.md';

const intl = {
  messages,
  confirmation,
  information,
};

export default intl;
