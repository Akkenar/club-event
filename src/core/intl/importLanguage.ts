export function importLanguage(language: string) {
  switch (language) {
    case 'en':
      return import(/* webpackChunkName: 'en' */ './en');
    case 'de':
      return import(/* webpackChunkName: 'de' */ './de');
    case 'fr':
      return import(/* webpackChunkName: 'fr' */ './fr');
    case 'it':
      return import(/* webpackChunkName: 'it' */ './it');
    default:
      throw new Error(`${language} not supported`);
  }
}
