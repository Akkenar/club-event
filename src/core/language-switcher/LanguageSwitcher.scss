@import url('../../../node_modules/semantic-ui-css/components/dropdown.min.css');
@import url('../../../node_modules/semantic-ui-css/components/transition.min.css');

.LanguageSwitcher .text {
  width: 2em;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}
