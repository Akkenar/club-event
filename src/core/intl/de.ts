import * as confirmation from '../../../translations/de/confirmation.md';
import * as information from '../../../translations/de/information.md';
import * as messages from '../../../translations/de/messages.json';
import { IntlType } from './intl.type';

const intl: IntlType = {
  confirmation,
  information,
  messages: (messages as any).default,
};

export default intl;
