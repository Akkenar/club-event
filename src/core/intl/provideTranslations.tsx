import * as React from 'react';
import {
  BrowserRouter,
  match as matchType,
  Redirect,
  Route,
  Switch,
} from 'react-router-dom';
import { getDefaultLanguage } from './getDefaultLanguage';
import LanguageContext from './LanguageContext';
import { useTranslations } from './useTranslations';

export const BASE_URL_WITH_LANG = '/:language([a-z]{2})';

interface ProvideMessagesProps {
  match: matchType<{ language: string }>;
}

export default function provideTranslations(
  Component: React.ComponentType<any>,
) {
  /**
   * The provided, depends on the routing and the language.
   * @param match
   * @constructor
   */
  const TranslationsProvider = ({ match }: ProvideMessagesProps) => {
    const targetLanguage = match.params.language;
    const value = useTranslations(targetLanguage);

    return (
      <LanguageContext.Provider value={value}>
        <Component />
      </LanguageContext.Provider>
    );
  };

  // Wrap the translation provider in a router.
  return () => (
    <BrowserRouter>
      <Switch>
        <Route
          path={`${BASE_URL_WITH_LANG}/:page`}
          component={TranslationsProvider}
        />
        <Redirect to={`/${getDefaultLanguage()}/home`} />
      </Switch>
    </BrowserRouter>
  );
}
