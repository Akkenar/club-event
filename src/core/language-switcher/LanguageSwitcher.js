import React, { useContext } from 'react';
import getKey from '../intl/getKey';
import { Dropdown } from 'semantic-ui-react';
import LanguageContext from '../intl/LanguageContext';
import { NavLink } from 'react-router-dom';

import './LanguageSwitcher.scss';

const LanguageSwitcher = () => {
  const { handleChangeLanguage, messages } = useContext(LanguageContext);

  return (
    <Dropdown
      className="link item right LanguageSwitcher"
      text={getKey('current.language', messages)}
    >
      <Dropdown.Menu>
        <Dropdown.Item
          title="FranÃ§ais"
          as={NavLink}
          to="/fr"
          onClick={() => handleChangeLanguage('fr')}
        >
          FR
        </Dropdown.Item>
        <Dropdown.Item
          title="Deutsch"
          as={NavLink}
          to="/de"
          onClick={() => handleChangeLanguage('de')}
        >
          DE
        </Dropdown.Item>
        <Dropdown.Item
          title="Italianno"
          as={NavLink}
          to="/it"
          onClick={() => handleChangeLanguage('it')}
        >
          IT
        </Dropdown.Item>
        <Dropdown.Item
          title="English"
          as={NavLink}
          to="/en"
          onClick={() => handleChangeLanguage('en')}
        >
          EN
        </Dropdown.Item>
      </Dropdown.Menu>
    </Dropdown>
  );
};

export default LanguageSwitcher;
