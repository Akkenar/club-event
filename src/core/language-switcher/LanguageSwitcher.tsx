import * as React from 'react';
import { useContext } from 'react';
import { Dropdown } from 'semantic-ui-react';
import getKey from '../intl/getKey';
import LanguageContext from '../intl/LanguageContext';

import './LanguageSwitcher.scss';

const Button = (props: any) => (
  <button
    {...props}
    className={`${props.className || ''} LanguageSwitcher__button`}
  />
);

const LanguageSwitcher = () => {
  const { handleChangeLanguage, messages } = useContext(LanguageContext);

  // So we don't use a Lambda in the jsx.
  const changeLanguage = (language: string) => () =>
    handleChangeLanguage(language);

  return (
    <Dropdown
      className="link item right LanguageSwitcher"
      text={getKey('current.language', messages)}
    >
      <Dropdown.Menu>
        <Dropdown.Item
          title="FranÃ§ais"
          as={Button}
          onClick={changeLanguage('fr')}
        >
          FR
        </Dropdown.Item>
        <Dropdown.Item
          title="Deutsch"
          as={Button}
          onClick={changeLanguage('de')}
        >
          DE
        </Dropdown.Item>
        <Dropdown.Item
          title="Italianno"
          as={Button}
          onClick={changeLanguage('it')}
        >
          IT
        </Dropdown.Item>
        <Dropdown.Item
          title="English"
          as={Button}
          onClick={changeLanguage('en')}
        >
          EN
        </Dropdown.Item>
      </Dropdown.Menu>
    </Dropdown>
  );
};

export default LanguageSwitcher;
