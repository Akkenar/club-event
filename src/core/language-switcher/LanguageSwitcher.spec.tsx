import * as React from 'react';
import { BrowserRouter, Redirect, Route, Switch } from 'react-router-dom';
import { render } from 'react-testing-library';
import { IntlType } from '../intl/intl.type';
import LanguageContext from '../intl/LanguageContext';
import { BASE_URL_WITH_LANG } from '../intl/provideTranslations';
import LanguageSwitcher from './LanguageSwitcher';

jest.mock('../intl/importLanguage', () => ({
  preloadAllLanguages: jest.fn(),
}));

import { preloadAllLanguages } from '../intl/importLanguage';

const context: IntlType = {
  confirmation: '',
  information: '',
  schedule: '',
  language: 'en',
  messages: {},
};

const LanguageSwitcherWithRouter = () => {
  const Component = () => (
    <LanguageContext.Provider value={context}>
      <LanguageSwitcher />
    </LanguageContext.Provider>
  );

  return (
    <BrowserRouter>
      <Switch>
        <Route path={`${BASE_URL_WITH_LANG}/:page`} component={Component} />
        <Redirect path="/" exact={true} to="/en/home" />
      </Switch>
    </BrowserRouter>
  );
};

describe('LanguageSwitcher', () => {
  it('should match snapshot', () => {
    const wrapper = render(<LanguageSwitcherWithRouter />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should preload all the languages when opening', () => {
    const wrapper = render(<LanguageSwitcherWithRouter />);
    expect(preloadAllLanguages).toHaveBeenCalledTimes(0);

    wrapper.getByTestId('language-switcher').click();
    expect(preloadAllLanguages).toHaveBeenCalledTimes(1);
  });
});
