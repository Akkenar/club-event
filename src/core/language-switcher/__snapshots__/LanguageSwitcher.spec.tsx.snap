// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSwitcher should match snapshot 1`] = `
<body>
  <div>
    <div
      aria-expanded="false"
      class="ui dropdown link item right LanguageSwitcher"
      role="listbox"
      tabindex="0"
    >
      <div
        aria-live="polite"
        class="text"
        role="alert"
      >
        current.language
      </div>
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="menu transition"
      >
        <button
          class="item LanguageSwitcher__button"
          role="option"
          title="FranÃ§ais"
        >
          FR
        </button>
        <button
          class="item LanguageSwitcher__button"
          role="option"
          title="Deutsch"
        >
          DE
        </button>
        <button
          class="item LanguageSwitcher__button"
          role="option"
          title="Italianno"
        >
          IT
        </button>
        <button
          class="item LanguageSwitcher__button"
          role="option"
          title="English"
        >
          EN
        </button>
      </div>
    </div>
  </div>
</body>
`;
